### PR TITLE
add `/debug/ecdsz`

### PR DESCRIPTION
**Pull request by zirain**
**Please provide a description of this PR:**

right now ecds is not a part of configdump, this PR make you can debug with endpoint `/debug/ecdsz` 

demo usage:

```console
kubectl exec -it istiod-7db4959cf9-btrj2 -nistio-system -- pilot-discovery request get /debug/ecdsz?proxyID=httpbin-74fb669cc6-mxhpx
```

```json
[
    {
        "config": {
            "name": "default.httpbin-rate-limiting",
            "Vm": {
                "VmConfig": {
                    "runtime": "envoy.wasm.runtime.v8",
                    "code": {
                        "Specifier": {
                            "Remote": {
                                "http_uri": {
                                    "uri": "oci://ghcr.io/tetratelabs/wasm-rate-limiting:v1",
                                    "HttpUpstreamType": {
                                        "Cluster": "_"
                                    },
                                    "timeout": {
                                        "seconds": 30
                                    }
                                }
                            }
                        }
                    }
                }
            },
            "configuration": {}
        }
    },
    {
        "config": {
            "name": "default.display-metadata",
            "Vm": {
                "VmConfig": {
                    "runtime": "envoy.wasm.runtime.v8",
                    "code": {
                        "Specifier": {
                            "Remote": {
                                "http_uri": {
                                    "uri": "oci://ghcr.io/zirain/display-metadata:latest",
                                    "HttpUpstreamType": {
                                        "Cluster": "_"
                                    },
                                    "timeout": {
                                        "seconds": 30
                                    }
                                }
                            }
                        }
                    },
                    "environment_variables": {
                        "host_env_keys": [
                            "POD_NAMESPACE"
                        ]
                    }
                }
            },
            "configuration": {}
        }
    }
]
```

@bianpengyuan  please take a first reviews.

### CODE REVIEW

**Review by bianpengyuan — COMMENTED** [review]
Can we decode and unmarshal if the type is Wasm extension? If it is other type, it would also be useful to get the type of the extension, although probably not possible to unmarshal. Also notice that we are using env at VM config to passing image pull secret, so will need some redaction at the dump.

**Comment by zirain** [comment]
> Can we decode and unmarshal if the type is Wasm extension? If it is other type, it would also be useful to get the type of the extension, although probably not possible to unmarshal. Also notice that we are using env at VM config to pass image pull secret, so will need some redaction at the dump.

I think we can decode and unmarshal, but maybe some perfermance issue when there're too much ecds configs.

BTW, I think `EcdsGenerator` do not respect the rule `If empty, all resources of the
 TypeUrl type are watched`, can you confirm that?
https://github.com/istio/istio/blob/56edd90bb930ff715474769e8c9587f8c74700eb/pilot/pkg/model/context.go#L320

**Comment by zirain** [comment]
@bianpengyuan updated with unmarshal

File: pilot/pkg/xds/debug.go [hunk 1]
```diff
@@ -580,6 +583,60 @@ func (s *DiscoveryServer) adsz(w http.ResponseWriter, req *http.Request) {
...
+	if s.handlePushRequest(w, req) {
+		return
+	}
+	proxyID, con := s.getDebugConnection(req)
+	if con == nil {
+		s.errorHandler(w, proxyID, con)
+		return
+	}
+
+	if s.Generators[v3.ExtensionConfigurationType] != nil {
```
Comment (howardjohn) [inline pilot/pkg/xds/debug.go:598]: maybe we should just make config_dump able to take in a list of generator keys. We can have shortcuts sa well. So something like `config_dump?types=ecds`.

Then we do not need a bunch of new methods for every type we add?

File: pilot/pkg/xds/debug.go [hunk 2]
```diff
@@ -580,6 +583,60 @@ func (s *DiscoveryServer) adsz(w http.ResponseWriter, req *http.Request) {
...
+}
+
+func unmarshalToWasm(r *discovery.Resource) (*wasm.Wasm, error) {
+	tce := &core.TypedExtensionConfig{}
+	if err := r.GetResource().UnmarshalTo(tce); err != nil {
+		return nil, err
+	}
+
+	w := &wasm.Wasm{}
+	if err := tce.TypedConfig.UnmarshalTo(w); err != nil {
```
Comment (bianpengyuan) [inline pilot/pkg/xds/debug.go:633]: I think user probably would use EnvoyFilter to add ECDS resource which is not of type Wasm? We probably should not warn, but instead just dump it as unmarshalled typedconfig?

**Comment by bianpengyuan** [comment]
> I think we can decode and unmarshal, but maybe some perfermance issue when there're too much ecds configs.

This is a debug interface, so performance is not much of a concern.

**Comment by howardjohn** [comment]
We should be careful to not leak Secrets once the pullSecret PR is merged

**Comment by bianpengyuan** [comment]
> We should be careful to not leak Secrets once the pullSecret PR is merged

Yeah, I added to the head of my priority list :).

File: pilot/pkg/xds/debug.go (hunk 1 above)
Comment (zirain) [reply pilot/pkg/xds/debug.go:598]: right now, `adsz` have something like what your said, `confug_dump` will dump everything .

https://github.com/istio/istio/blob/e3904269b7bd95207fdccd2be6f14276a3ad470a/pilot/pkg/xds/debug.go#L177

https://github.com/istio/istio/blob/e3904269b7bd95207fdccd2be6f14276a3ad470a/pilot/pkg/xds/debug.go#L195

I'm not sure that `config_dump?types=ecds` is a good idea or not.

File: pilot/pkg/xds/debug.go (hunk 2 above)
Comment (zirain) [reply pilot/pkg/xds/debug.go]: updated

File: pilot/pkg/xds/debug.go [hunk 3]
```diff
@@ -580,6 +583,69 @@ func (s *DiscoveryServer) adsz(w http.ResponseWriter, req *http.Request) {
...
+			}
+		}
+
+		writeJSON(w, wasmCfgs)
+	}
+}
+
+const (
+	apiTypePrefix      = "type.googleapis.com/"
+	wasmHTTPFilterType = apiTypePrefix + "envoy.extensions.filters.http.wasm.v3.Wasm"
```
Comment (bianpengyuan) [inline pilot/pkg/xds/debug.go:628]: Looks like these const are repeated at several places. Can we consolidate them into model.go?

File: pilot/pkg/xds/debug.go [hunk 4]
```diff
@@ -580,6 +583,69 @@ func (s *DiscoveryServer) adsz(w http.ResponseWriter, req *http.Request) {
...
+	if con == nil {
+		s.errorHandler(w, proxyID, con)
+		return
+	}
+
+	if s.Generators[v3.ExtensionConfigurationType] != nil {
+		r, ok := con.proxy.WatchedResources[v3.ExtensionConfigurationType]
+		if !ok {
+			w.WriteHeader(http.StatusNotFound)
+			_, _ = w.Write([]byte(fmt.Sprintf("not watched ExtensionConfigurationType found, proxyID: %s\n", proxyID)))
```
Comment (bianpengyuan) [inline pilot/pkg/xds/debug.go:602]:
```suggestion
			_, _ = w.Write([]byte(fmt.Sprintf("no watched ExtensionConfigurationType found, proxyID: %s\n", proxyID)))
```

File: pilot/pkg/xds/debug.go (hunk 3 above)
Comment (zirain) [reply pilot/pkg/xds/debug.go:628]: actually yes, but I want refactor all this in a another single PR.

Comment (zirain) [reply pilot/pkg/xds/debug.go:628]: there's something same in `pkg/wasm/convert.go`, we'd find a better place where we can share these const.

File: pilot/pkg/xds/debug.go (hunk 1 above)
Comment (zirain) [reply pilot/pkg/xds/debug.go:598]: tracking in https://github.com/istio/istio/issues/37504

**Review by bianpengyuan — APPROVED** [review]

**Pull request merged**